components/raftstore/src/store/worker/region.rs

```diff
@@ -627,6 +642,46 @@ where
 
         Ok(())
     }
+
+    fn pre_apply_snapshot(&self, task: &Task<EK::Snapshot>) -> Result<()> {
+        let (region_id, abort, peer_id) = match task {
+            Task::Apply {
+                region_id,
+                status,
+                peer_id,
+            } => (region_id, status.clone(), peer_id),
+            _ => panic!("invalid apply snapshot task"),
+        };
+
+        let region_state = self.get_region_state(*region_id)?;
+        let apply_state = self.get_apply_state(*region_id)?;
+
+        let timer = Instant::now();
+        check_abort(&abort)?;
+
+        let term = apply_state.get_truncated_state().get_term();
+        let idx = apply_state.get_truncated_state().get_index();
+        let snap_key = SnapKey::new(*region_id, term, idx);
+        let s = box_try!(self.mgr.get_snapshot_for_applying(&snap_key));
+        if !s.exists() {
+            return Err(box_err!("missing snapshot file {}", s.path()));
+        }
+        check_abort(&abort)?;
+        self.coprocessor_host.pre_apply_snapshot(
```

Comment:
Why not put it in `apply_snap`? So all these states and snapshot don't need to be fetched again.

Reply:
If we do `pre_apply_snapshot` in `apply_snap`, which means eventually popped from pending task queue by `handle_pending_applies`, that will be too late. 
TiFlash need this because we can do `pre_apply_snapshot` for multiple snapshots in parallel.
If this procedure brings extra costs, I think I can introduce a feature `need_pre_apply_snapshot` in engine_trait as a switch.